Objects/moduleobject.c

```diff
@@ -559,65 +559,88 @@ PyModule_GetState(PyObject* m)
 void
 _PyModule_Clear(PyObject *m)
 {
-    PyObject *d = ((PyModuleObject *)m)->md_dict;
-    if (d != NULL)
-        _PyModule_ClearDict(d);
+    PyModuleObject *module = (PyModuleObject *)m;
+    _PyModule_ClearDict(module->md_name, module->md_dict);
 }
 
 void
-_PyModule_ClearDict(PyObject *d)
+_PyModule_ClearDict(PyObject *module_name, PyObject *dict)
 {
+    int verbose = _Py_GetConfig()->verbose;
+    if (verbose) {
+        PySys_FormatStderr("# cleanup[3] wiping %S module\n", module_name);
+    }
+
+    // If the module has no dict: there is nothing to do.
+    if (dict == NULL) {
+        return;
+    }
+
     /* To make the execution order of destructors for global
        objects a bit more predictable, we first zap all objects
        whose name starts with a single underscore, before we clear
        the entire dictionary.  We zap them by replacing them with
        None, rather than deleting them from the dictionary, to
        avoid rehashing the dictionary (to some extent). */
 
-    Py_ssize_t pos;
-    PyObject *key, *value;
-
-    int verbose = _Py_GetConfig()->verbose;
+    for (int step=1; step <= 2; step++) {
+        PyObject *reversed = PyObject_CallOneArg((PyObject*)&PyReversed_Type, dict);
+        if (reversed == NULL) {
+            PyErr_WriteUnraisable(NULL);
+            return;
+        }
+        PyObject *keys = PyObject_CallOneArg((PyObject*)&PyList_Type, reversed);
+        Py_DECREF(reversed);
+        if (keys == NULL) {
+            PyErr_WriteUnraisable(NULL);
+            return;
+        }
+        PyObject *iter = PyObject_GetIter(keys);
```

Comment:
It is easier to iterate list by index.

Reply:
Oh, you're right. My first attempt avoided the creation of a list, but then I get annoying RuntimeError when the dict was modified while we iterate on it. Deleting a variable can have many side effects, it's way safer to copy the list of keys.